### ISSUE TITLE
Nested flex-box layouts don't respect justify-content tag

### ISSUE THREAD

**Issue by tokyovigilante** [Kozea/WeasyPrint#673]
A final further flex-box bug I've discovered as above. I'm trying to position two columns of text with flex  containers, then within the right column, use a nested flex container to display a left-aligned and right aligned heading on a single row using `{ justify-content: space-around; }`, for example:
```
|--------|-----------------------------------|
| Col 1  | Left-align            Right-align |
|        | Body text block follows...        |
|--------|-----------------------------------|
```

<details><summary>
Sample un-nested code (works as expected):
</summary>

```html
<!DOCTYPE html>
<html>
<meta charset="UTF-8">
    <title></title>
    <style type="text/css">
        @media print {
                    @page {
                        size: A4 portrait;
                        margin: 1.5cm !important;
                        margin-top: 1cm !important;
                    }
                }
                body {
                    font-family: "serif";
                    font-size: 11pt !important;
                    text-align: left;
                    line-height: 1.25;
                }
                .d-flex {
                    display: flex !important;
                }
                .col1 {
                    flex: 0 0 90pt;
                }
                .col2 {
                    width: 100%;
                }
                .d-flex {
                    display: flex !important;
                }
                .justify-between {
                    justify-content: space-between;
                }
            }
        </style>
</head>
<body>
    <div class="col2 d-flex justify-between">
        <p><strong>Left-aligned text</strong></p>
        <p><em>Right-aligned text</em></p>
    </div>
</body>
</html>

```

</details>

<details><summary>
Sample nested code:
</summary>

```html

<!DOCTYPE html>
<html>
<meta charset="UTF-8">
    <title></title>
    <style type="text/css">
        @media print {
                    @page {
                        size: A4 portrait;
                        margin: 1.5cm !important;
                        margin-top: 1cm !important;
                    }
                }
                body {
                    font-family: "serif";
                    font-size: 11pt !important;
                    text-align: left;
                    line-height: 1.25;
                }
                .d-flex {
                    display: flex !important;
                }
                .col1 {
                    flex: 0 0 90pt;
                }
                .col2 {
                    flex: auto;
                }
                .d-flex {
                    display: flex !important;
                }
                .justify-between {
                    justify-content: space-between;
                }
            }
        </style>
</head>
<body>
    <div class="d-flex">
        <div class="col1">
            <p><strong>Section 1</strong></p>
        </div>
        <div class="col2 d-flex justify-between">
            <p><strong>Left-aligned text</strong></p>
            <p><em>Right-aligned text</em></p>
        </div>
    </div>
</body>
</html>

```

</details>

Final bonus bug: [`{ justify-content: space-evenly; }`](https://drafts.csswg.org/css-align-3/#typedef-content-distribution) and `{ justify-content: stretch; }` (which I'd love to be able to use) don't seem to be implemented: `WARNING: Ignored `justify-content: space-evenly` at 56:5, invalid value.`

**Comment by liZe** [comment]
I'm not sure about the fix, but I'll check that while closing #601 :wink:.

**Comment by tokyovigilante** [comment]
Hi, thanks again, this fix works well but for bug #672 :)

**Comment by liZe** [comment]
> Hi, thanks again, this fix works well but for bug #672 :)

Oops, sorry!

**Comment by liZe** [comment]
There were so many small fixes to solve this problem… Thanks a lot for providing real-life examples, it's really useful.

> Final bonus bug: { justify-content: space-evenly; } and { justify-content: stretch; }

Oh, I've seen these values in the W3C test suite and wondered why the values weren't in the flexbox spec. It's yet another working draft, of course! It should be pretty easy to implement, I'll close the issue when it's ready.

**Comment by liZe** [comment]
It's time for #601, it's almost impossible to change a line in `flex.py` without breaking everything.

**Comment by tokyovigilante** [comment]
Thanks again for adding the extra `justify-content` flags, however 6f4bb9a only seems to refer to the missing flags rather than actually addressing the main bug which was about being able to use nested flex layouts.

On a side note regarding #601 I'm happy to start chipping away at these. Do I basically need to follow the existing test format and import the test HTML from the relevant spec?

**Comment by liZe** [comment]
> Thanks again for adding the extra justify-content flags, however 6f4bb9a only seems to refer to the missing flags rather than actually addressing the main bug which was about being able to use nested flex layouts.

Yes, I've not added the issue number in the related commits, but it's actually a sum of little independent fixes (63e11ec0, 7ab39313, 2c7caf5f, 674126f7, 64a6bf72) that are more or less related to your problem.

> On a side note regarding #601 I'm happy to start chipping away at these. Do I basically need to follow the existing test format and import the test HTML from the relevant spec?

I would love to have some help on #601! I've added an empty `weasyprint/tests/test_layout/test_flex.py` file where you can add tests. We generally don't copy HTML samples from the spec or from the W3C test suite (it's generally better to have even smaller, specifically crafted tests), but it is an obvious source of inspiration.

(The next paragraphs are not 100% useful if you know how Python tests work.)

You can take a look at other files in `test_layout` to know how test work (`test_block.py` is a good way to start). If you test 100% valid syntaxes (and you probably will), you can add the `@assert_no_logs`decorator to check that no error occurred during the test.

You can also check the `@pytest.mark.parametrize` decorator if you have to test many similar use cases.

Tests can be launched using the `./setup.py test` command line, it will install everything you need in a dedicated environment. You can add pytest options if you want: using `./setup.py test --addopts='--no-cov -k test_flex.py'` will for example launch tests only in the `test_flex.py` file with no coverage.
